src/audio/source_api_helper.c

```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2023 Intel Corporation. All rights reserved.
+//
+
+#include <sof/audio/source_api.h>
+#include <sof/audio/source_api_implementation.h>
+#include <sof/audio/audio_stream.h>
+
+void source_init(struct sof_source __sparse_cache *source, const struct source_ops *ops,
+		 struct sof_audio_stream_params *audio_stream_params)
+{
+	source->ops = ops;
+	source->requested_read_frag_size = 0;
+	source->audio_stream_params = audio_stream_params;
```

Comment:
@marcinszkudlinski in fact this is a bug. You call this function with an *acquired* object and you store a pointer to that acquired object permanently and then you use it outside of the object ownership scope. This will cause failures. Please fix in a follow up.

Reply:
In case of audio_stream handlers to sink/sources are __sparse
```
static inline struct sof_sink __sparse_cache *
audio_stream_get_sink(struct audio_stream __sparse_cache *audio_stream)
{
    return &audio_stream->sink_api;
}

```

That implies that the handlers may be used when buffer is aquired, and only there
**This is enforced by module_adapter and there's no danger here**

The only concern is academic - we keep something that should not be used without lock. 
At the moment we may think about shadowing audio_stream_params  in sink/source structure as a temporary sollution (till comp_buffer/audio stream is still in use), this will cost some RAM. 

Next implementation (dp_queue) will keep audio_stream_params in shared, never cached memory - no shadowing needed. Other upcoming implementations - like shared LL buffer - will be used exclusively by single core (so may be safely always cached - no shadowing needed).